### PR TITLE
rados: implement rados_getaddrs

### PR DESCRIPTION
**Pull request by gman0**
This PR implements binding for rados_getaddrs.

Includes a unit test.

## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code
- [x] Is this a new API? Added a new file that begins with `//go:build ceph_preview`
- [x] Ran `make api-update` to record new APIs

### CODE REVIEW

File: rados/rados_getaddrs_test.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+//go:build ceph_preview
+// +build ceph_preview
```
Comment (anoopcs9) [inline rados/rados_getaddrs_test.go:2]: I think we can safely ignore adhering to the backward compatible variant these days.

File: rados/rados_getaddrs.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+//go:build ceph_preview
+// +build ceph_preview
```
Comment (anoopcs9) [inline rados/rados_getaddrs.go:2]: I think we can safely ignore adhering to the backward compatible variant these days.

**Review by anoopcs9 — COMMENTED** [review]
See below for a small comment.

File: rados/rados_getaddrs.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// #include <stdlib.h>
+// #include <rados/librados.h>
+import "C"
+
+import (
+	"unsafe"
+)
+
+// GetAddrs wraps the function to get the addresses of
+// the RADOS session, suitable for blocklisting.
```
Comment (phlogistonjohn) [inline rados/rados_getaddrs.go:16]: Let's tell users what the function does not what it wraps. The Implements section handles telling them what it wraps, so:
```
GetAddrs returns a string containing the addresses of the RADOS session, suitable for blocklisting.
```
or similar.

File: rados/rados_getaddrs.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+//
+// Implements:
+//
+//	int rados_getaddrs(rados_t cluster, char **addrs)
+func (c *Conn) GetAddrs() (string, error) {
```
Comment (phlogistonjohn) [inline rados/rados_getaddrs.go:21]: I was curious about the formatting of this string. I looked in the python and C header file in the ceph tree but nothing documents the format of this string. So I suppose it's meant to be opaque. That's OK.  If we ever needed to parse it in the future we can add a helper function for that later.

File: rados/rados_getaddrs_test.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import (
+	"github.com/stretchr/testify/assert"
+)
+
+func (suite *RadosTestSuite) TestGetAddrs() {
+	suite.SetupConnection()
+
+	addrs, err := suite.conn.GetAddrs()
+	assert.NoError(suite.T(), err)
+	assert.NotEmpty(suite.T(), addrs, "rados_getaddrs")
```
Comment (phlogistonjohn) [inline rados/rados_getaddrs_test.go:15]: I suppose there's nothing else we can assert about the content of this string, right?

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
One request, and a pair of thoughts.

File: rados/rados_getaddrs.go (hunk 4 above)
Comment (gman0) [reply rados/rados_getaddrs.go:21]: I was also wondering about that, but if librados doesn't bother parsing the output into something manageable, then I guess we can treat it the same way?

* [`rados_getaddrs()`](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/librados/librados_c.cc#L526-L535) calls `librados::RadosClient::get_addrs()`
* [`librados::RadosClient::get_addrs()`](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/librados/RadosClient.cc#L777-L781) calls `*cos << messenger->get_myaddrs();`
  * [`get_myaddrs()`](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/msg/Messenger.h#L221-L229) returns [`entity_addrvec_t`](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/msg/msg_types.h#L581-L750)
    * [`friend std::ostream& operator<<(std::ostream& out, const entity_addrvec_t& av)`](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/msg/msg_types.h#L728-L736) -- here, if the vector has only a single item it prints only that one entity address:
      * Using this function [`std::ostream& operator<<(std::ostream& out, const entity_addr_t &addr)`](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/msg/msg_types.cc#L220-L230)
    * If the vector has more items, it uses formats it into `[item1,item2,...]` using [`inline std::ostream& operator<<(std::ostream& out, const std::vector<A,Alloc>& v)`](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/include/types.h#L134-L145)

Let me know if you'd like me also to parse the output in some way.

Comment (gman0) [reply rados/rados_getaddrs.go:21]: So far I got only a single entity address, so there is no weird formatting and only the `<IP>:<port>/nonce` is returned, but I guess there can be more.

File: rados/rados_getaddrs_test.go (hunk 5 above)
Comment (gman0) [reply rados/rados_getaddrs_test.go:15]: Most likely not, unless we parse the string. But if so, then that should be done by `GetAddrs` itself?

File: rados/rados_getaddrs.go (hunk 4 above)
Comment (phlogistonjohn) [reply rados/rados_getaddrs.go:21]: OK, let's assume the content is informational and/or opaque.

File: rados/rados_getaddrs_test.go (hunk 5 above)
Comment (phlogistonjohn) [reply rados/rados_getaddrs_test.go:15]: Agreed, as noted in the other comment let's assume the string is opaque.

And... that's fine. It was somewhat of a rhetorical question... but I felt the need to write it anyway. :-)

**Comment by gman0** [comment]
@phlogistonjohn I've fixed the comment. @anoopcs9 so I should remove the preview build tag? In the docs it says all new APIs must have it.

**Review by anoopcs9 — DISMISSED** [review]

File: docs/api-status.json [hunk 6]
```diff
@@ -1159,6 +1159,12 @@
         "comment": "Exec executes an OSD class method on an object.\nSee rados_exec() in the RADOS C api documentation for a general description.\n\nImplements:\n\n\tvoid rados_write_op_exec(rados_write_op_t write_op,\n\t                         const char *cls,\n\t                         const char *method,\n\t                         const char *in_buf,\n\t                         size_t in_len,\n\t                         int *prval)\n",
         "added_in_version": "v0.29.0",
         "expected_stable_version": "v0.31.0"
+      },
+      {
+        "name": "Conn.GetAddrs",
+        "comment": "GetAddrs wraps the function to get the addresses of\nthe RADOS session, suitable for blocklisting.\n\nImplements:\n\n\tint rados_getaddrs(rados_t cluster, char **addrs)\n",
```
Comment (phlogistonjohn) [inline docs/api-status.json:1165]: This generated JSON needs to be updated now that the code comment is changed. Try using git to revert the change to *just this file* and then re-run the make command to regenerate the JSON. (then squash it all back together :-) )

Comment (phlogistonjohn) [reply docs/api-status.json:1165]: If you need more instruction on all that feel free to ask.

**Comment by anoopcs9** [comment]
There seems to be a newer version of _protobuf_(and _protobuf-compiler_) causing update issues with `squid` CI jobs. This is due to the fact that _protobuf-compiler_ is only available with CRB repository which is not enabled by default with ceph container images. Let's see if we could get another rebuild of `squid` with the latest _protobuf_ and _protobuf-compiler_ in time.

**Comment by phlogistonjohn** [comment]
> There seems to be a newer version of _protobuf_(and _protobuf-compiler_) causing update issues with `squid` CI jobs. This is due to the fact that _protobuf-compiler_ is only available with CRB repository which is not enabled by default with ceph container images. Let's see if we could get another rebuild of `squid` with the latest _protobuf_ and _protobuf-compiler_ in time.

It's not a required check so we have the option of trying to fix it and then asking @gman0 to rebase or just ignoring the failure for this PR. I will look into fixing it now... but if it proves tricky we can skip it.

**Comment by phlogistonjohn** [comment]
squid workaround in #1038

**Review by anoopcs9 — APPROVED** [review]
lgtm.

**Comment by anoopcs9** [comment]
@Mergifyio rebase

**Review by phlogistonjohn — APPROVED** [review]
lgtm
